src/main/java/org/apache/sysds/runtime/compress/colgroup/ColGroupLinearFunctional.java

```diff
@@ -0,0 +1,516 @@
+
+			double[] colSumsAndWeightedColSums = new double[2 * lhs.getNumCols()];
+			for(int j = 0, offTmp = 0; j < lhs.getNumCols(); j++, offTmp += 2) {
+				for(int i = 0; i < numRowsLeft; i++) {
+					colSumsAndWeightedColSums[offTmp] += lhs.getIdx(i, j);
+					colSumsAndWeightedColSums[offTmp + 1] += (i+1) * lhs.getIdx(i, j);
+				}
+			}
+
+			MatrixBlock sumMatrix = new MatrixBlock(lhs.getNumCols(), 2, colSumsAndWeightedColSums);
+			MatrixBlock coefficientMatrix = DataConverter.convertToMatrixBlock(_coefficents);
+
+			LibMatrixMult.matrixMult(sumMatrix, coefficientMatrix, tmpRet);
+		} else if(lhs instanceof ColGroupLinearFunctional) {
+			ColGroupLinearFunctional lhsLF = (ColGroupLinearFunctional) lhs;
+
+			final double sumIndices = _numRows * (_numRows + 1)/2.0;
+			final double sumSquaredIndices = _numRows * (_numRows + 1) * (2*_numRows + 1)/6.0;
+
+			MatrixBlock weightMatrix = new MatrixBlock(2, 2, new double[] {_numRows, sumIndices, sumIndices, sumSquaredIndices});
+			MatrixBlock coefficientMatrixLhs = DataConverter.convertToMatrixBlock(lhsLF._coefficents);
+			MatrixBlock coefficientMatrixRhs = DataConverter.convertToMatrixBlock(_coefficents);
+
+			coefficientMatrixLhs = LibMatrixReorg.transposeInPlace(coefficientMatrixLhs,
+				InfrastructureAnalyzer.getLocalParallelism());
+
+			// We simply compute a matrix multiplication chain in coefficient space, i.e.,
+			// 					t(L) %*% R = t(coeff(L)) %*% W %*% coeff(R)
+			// where W is a weight matrix capturing the size of the shared dimension (weightMatrix above)
+			// and coeff(X) denotes the 2 x n matrix of the m x n matrix X.
+			MatrixBlock tmp = new MatrixBlock(lhs.getNumCols(), 2, false);
+			LibMatrixMult.matrixMult(coefficientMatrixLhs, weightMatrix, tmp);
+			LibMatrixMult.matrixMult(tmp, coefficientMatrixRhs, tmpRet);
+		} else if(lhs instanceof APreAgg) {
+			// TODO: implement
+			throw new NotImplementedException();
+		} else {
+			throw new NotImplementedException();
+		}
+
+		final double[] resV = result.getDenseBlockValues();
+		if(tmpRet.isEmpty())
+			return;
+		else if(tmpRet.isInSparseFormat()) {
+			SparseBlock sb = tmpRet.getSparseBlock();
+			for(int row = 0; row < lhs._colIndexes.length; row++) {
+				if(sb.isEmpty(row))
+					continue;
+				final int apos = sb.pos(row);
+				final int alen = sb.size(row) + apos;
+				final int[] aix = sb.indexes(row);
+				final double[] avals = sb.values(row);
+				final int offRes = lhs._colIndexes[row] * result.getNumColumns();
+				for(int col = apos; col < alen; col++)
+					resV[offRes + _colIndexes[aix[col]]] += avals[col];
+			}
+		}
+		else {
+			double[] tmpRetV = tmpRet.getDenseBlockValues();
+			for(int row = 0; row < lhs.getNumCols(); row++) {
+				final int offRes = lhs._colIndexes[row] * result.getNumColumns();
+				final int offTmp = row * getNumCols();
+				for(int col = 0; col < getNumCols(); col++) {
+					resV[offRes + _colIndexes[col]] += tmpRetV[offTmp + col];
+				}
+			}
+		}
```

Comment:
this code looks suspiciously much like the code inside ColGroupUncompressed,
i suggest that we combine the two into a util function in ColGroupUtil.